Variational bayes speaker diarization

saikiranvalluri opened the pull request:
VB resegmentation converts DER from baseline - 8.13% to 6.21%  after x-vector stage.

KarelVesely84 committed:
[src] Fix 'sausage-time' issue which occurs with disabled MBR decoding. (#2996)

aarora8 committed:
[egs] Add scripts for yomdle Russian (OCR task) (#2953)

NagendraKumarGoel committed:
[egs] Simplify lexicon preparation in Fisher callhome Spanish (#2999)

aarora8 committed:
[egs] Update GALE Arabic recipe (#2934)

aarora8 committed:
[egs] Remove outdated NN results from Gale Arabic recipe (#3002)

huangruizhe committed:
[egs] Add RESULTS file for the tedlium s5_r3 (release 3) setup (#3003)

danpovey committed:
[src] Fixes to grammar-fst code to handle LM-disambig symbols properly (#3000)

thanks: armando.muscariello@gmail.com

boeddeker committed:
[src] Cosmetic change to mel computation (fix option string) (#3011)

daanzu committed:
[src] Fix Visual Studio error due to alternate syntactic form of noreturn (#3018)

jybaek committed:
[egs] Fix location of sequitur installation (#3017)

daanzu committed:
[src] Fix w/ ifdef Visual Studio error from alternate syntactic form noreturn (#3020)

danpovey committed:
[egs] Some fixes to getting data in heroico recipe (#3021)

jtrmal committed:
[egs] BABEL script fix: avoid make_L_align.sh generating invalid files (#3022)

jdieguez committed:
[src] Fix to older online decoding code in online/ (OnlineFeInput; was broken by commit cc2469e8). (#3025)

oplatek committed:
[script] Fix unset bash variable in make_mfcc.sh (#3030)

oplatek committed:
[scripts]  Extend limit_num_gpus.sh to support --num-gpus 0.  (#3027)

Commit:
[scripts] fix bug in utils/add_lex_disambig.pl when sil-probs and pron-probs used (#3033)

bug would likely have resulted in determinization failure (only when not using word-position-dependent phones).

francoishernandez committed:
[egs] Fix path in Tedlium r3 rnnlm training script (#3039)

danpovey committed:
[src] Thread-safety for GrammarFst (thx:armando.muscariello@gmail.com) (#3040)

Commit:
[scripts] Cosmetic fix to get_degs.sh (#3045)

ChunChiehChang committed:
[egs] Small bug fixes for IAM and UW3 recipes (#3048)

danpovey committed:
[scripts] Nnet3 segmentation: fix default params (#3051)

igrinis committed:
[scripts] Allow perturb_data_dir_speed.sh to work with utt2lang (#3055)

xiaohui-zhang committed:
[scripts] Make beam in monophone training configurable  (#3057)

rezame committed:
[scripts] Allow reverberate_data_dir.py to support unicode filenames (#3060)

vimalmanohar committed:
[scripts] Make some cleanup scripts work with python3 (#3054)

jfainberg committed:
[scripts] bug fix to nnet2->3 conversion, fixes #886 (#3071)

luitjens committed:
[src] Make copies occur in per-thread default stream (for GPUs)  (#3068)

luitjens committed:
[src] Add GPU version of MergeTaskOutput().. relates to batch decoding (#3067)

luitjens committed:
[src] Add device options to enable tensor core math mode. (#3066)

kkm000 committed:
[src] Log nnet3 computation to VLOG, not std::cout (#3072)

danpovey committed:
[src] Allow upsampling in compute-mfcc-feats, etc. (#3014)

keli78 committed:
[src] fix problem with rand_r being undefined on Android (#3037)

Commit:
[egs] Update swbd1_map_words.pl, fix them_1's -> them's (#3052)

kkm000 committed:
[src] Add const overload OnlineNnet2FeaturePipeline::IvectorFeature (#3073)

antonstakhouski committed:
[src] Fix syntax error in egs/bn_music_speech/v1/local/make_musan.py (#3074)

pzelasko committed:
[src] Memory optimization for online feature extraction of long recordings (#3038)

kan-bayashi committed:
[build] fixed a bug in linux_configure_redhat_fat when use_cuda=no (#3075)

hhadian committed:
[scripts] Add missing '. ./path.sh' to get_utt2num_frames.sh (#3076)

hhadian committed:
[src,scripts,egs] Add count-based biphone tree tying for flat-start chain training (#3007)

desh2608 committed:
[scripts,egs] Remove sed from various scripts (avoid compatibility problems)  (#2981)

kkm000 committed:
[src] Rework error logging for safety and cleanliness (#3064)

desh2608 committed:
[src] Change warp-synchronous to cub::BlockReduce (safer but slower) (#3080)

kkm000 committed:
[src] Fix && and || uses where & and | intended, and other weird errors (#3087)

kkm000 committed:
[build] Some fixes to Makefiles (#3088)

clang is unhappy with '-rdynamic' in compile-only step, and the
switch is really unnecessary.

Also, the default location for MKL 64-bit libraries is intel64/.
The em64t/ was explained already obsolete by an Intel rep in 2010:
https://software.intel.com/en-us/forums/intel-math-kernel-library/topic/285973

pzelasko committed:
[src] Fixed -Wreordered warnings in feat (#3090)

entn-at committed:
[egs] Replace bc with perl -e (#3093)

nikhilm16 committed:
[scripts] Fix python3 compatibility issue in data-perturbing script (#3084)

csukuangfj committed:
[doc] fix some typos in doc. (#3097)

kkm000 committed:
[build] Make sure expf() speed probe times sensibly (#3089)

XIAOYixuan committed:
[scripts] Make sure merge_targets.py works in python3 (#3094)

desh2608 committed:
[src] ifdef to fix compilation failure on CUDA 8 and earlier (#3103)

csukuangfj committed:
[doc] fix typos and broken links in doc. (#3102)

freewym committed:
[scripts] Fix frame_shift bug in egs/swbd/s5c/local/score_sclite_conf.sh (#3104)

MartinKocour committed:
[src] Fix wrong assertion failure in nnet3-am-compute (#3106)

danpovey committed:
[src] Cosmetic changes to natural-gradient code (#3108)

KarelVesely84 committed:
[src,scripts] Python2 compatibility fixes and code cleanup for nnet1 (#3113)

KarelVesely84 committed:
[doc] Small documentation fixes; update on Kaldi history (#3031)

danpovey committed:
[src] Various mostly-cosmetic changes (copying from another branch) (#3109)

saikiranvalluri committed:
[scripts]  Simplify text encoding in RNNLM scripts (now only support utf-8) (#3065)

yfliao committed:
[egs] Add "formosa_speech" recipe (Taiwanese Mandarin ASR) (#2474)

rickychanhoyin committed:
[egs] python3 compatibility in csj example script (#3123)

danpovey committed:
[egs] python3 compatibility in example scripts (#3126)

psmit committed:
[scripts] Bug-fix for removing deleted words (#3116)

The type of --max-deleted-words-kept-when-merging in segment_ctm_edits.py
was a string, which prevented the mechanism from working altogether.

vimalmanohar committed:
[scripts] Add fix regarding num-jobs for segment_long_utterances*.sh(#3130)

jtrmal committed:
[src] Enable allow_{upsample,downsample} with online features (#3139)

Commit:
[src] Fix bad assert in fstmakecontextsyms (#3142)

daanzu committed:
[src] Fix to "Fixes to grammar-fst & LM-disambig symbols" (#3000) (#3143)

jtrmal committed:
[build] Make sure PaUtils exported from portaudio (#3144)

KarelVesely84 committed:
[src] cudamatrix: fixing a synchronization bug in 'normalize-per-row' (#3145)

was only apparent using large matrices

csukuangfj committed:
[src] Fix typo in comment (#3147)

danijel3 committed:
[src] Add binary that functions as a TCP server (#2938)

Shujian2015 committed:
[scripts] Fix bug in comment (#3152)

saikiranvalluri committed:
VB cleaning based diarization

saikiranvalluri committed:
[scripts] Fix bug in steps/segmentation/ali_to_targets.sh (#3155)

kkm000 committed:
[scripts] Avoid holding out more data than the requested num-utts (due to utt2uniq) (#3141)

dogancan committed:
[src,scripts] Add support for two-pass agglomerative clustering. (#3058)

kkm000 committed:
[src] Disable unget warning in PeekToken (and other small fix) (#3163)

btiplitz committed:
[build] Add new nvidia tools to windows build (#3159)

danijel3 committed:
[doc] Fix documentation errors and add more docs for tcp-server decoder  (#3164)

desh2608 committed:
[scripts] Fix non-randomness in getting utt2uniq, introduced in #3142 (#3175)

luitjens committed:
[build] Don't build for Tegra sm_XX versions on x86/ppc and vice versa; allow --cuda-arch overrides to have multiple versions (#3171)

aarora8 committed:
[egs] Fixes Re encoding to IAM, uw3 recipes (#3012)

luitjens committed:
[src] Efficiency improvement and extra checking for cudamarix, RE default stream  (#3182)

Shujian2015 committed:
[egs] Fix small typo in tedlium download script (#3178)

Commit:
[github] Add GitHub issue templates (#3187)

danpovey committed:
[build] Add missing dependency to Makefile (#3191)

hainan-xv committed:
[src] Fix bug in pruned lattice rescoring when input lattice has epsilons (#3190)

armusc committed:
[scripts] Fix bug in extend_lang.sh regarding extra_disambig.txt (#3195)

saikiranvalluri committed:
Fixes in callhome data prep and mem options

saikiranvalluri committed:
Appended VB resegmentation to v2 recipe

saikiranvalluri commented:
> Why not VB_root? Why VB_HMM_ROOT?

I think we are using speaker based HMMs in VB resegmentation.

saikiranvalluri committed:
Remove VB_clean folder

NagendraKumarGoel commented on egs/callhome_diarization/v1/local/dump_model.py:
```diff
@@ -0,0 +1,127 @@
+#!/home/hzili1/tools/anaconda3/envs/py27/bin/python
```
/usr/local/bin/env python

saikiranvalluri replied:
Sorry, resolved it.

saikiranvalluri committed:
Small bug fix in python version

NagendraKumarGoel commented on egs/callhome_diarization/v2/run.sh:
```diff
@@ -356,3 +368,83 @@ if [ $stage -le 11 ]; then
 fi
+
+if [ $VB_resegmentation ]; then
+# Variational Bayes method for smoothing the Speaker segments at frame-level
```
It should check here if VB tools have been downloaded and installed. Otherwise print the instructions for that.

saikiranvalluri replied:
I actually included that verification part in VB_resegmentation.sh

NagendraKumarGoel commented on tools/extras/install_diarization_VBHMM.sh:
```diff
@@ -0,0 +1,21 @@
+  unzip VB_diarization.zip
+fi
+
+pip install numexpr
```
does pip need --user argument? Just checking.

saikiranvalluri replied:
No, I just installed it in /usr/local/python lib, using the above cmd.

NagendraKumarGoel commented on egs/callhome_diarization/v1/local/VB_resegmentation.py:
```diff
@@ -0,0 +1,308 @@
+#!/usr/bin/env python
+
+import numpy as np
+import VB_diarization
```
this should be importing VB_diarization from BUT preferably.

saikiranvalluri replied:
Yes, it does, since we export PYTHONPATH in VB_resegmentation.sh

saikiranvalluri replied:
I verified that it is importing VB_diarisation module from BUT download.

jyhnnhyj committed:
[egs] Update Tedlium s5_r3 example with more up-to-date chain TDNN configuration

lucas-jo committed:
[scripts] Fix bug in extend_lang.sh causing validation failure w/ extra_disambig.txt (#3202)

saikiranvalluri committed:
Data preparation fixes for VB resegmentation

armusc committed:
[scripts] Bug-fix in make_lexicon_fst.py, which failed when --sil-prob=0  (#3206)

Shujian2015 committed:
[egs] Fix very small typo in run_tdnn_1b.sh (#3207)

langep committed:
[build] Tensorflow version update (#3204)

kangshiyin committed:
[src] Optimizations to CUDA kernels (#3209)

luitjens committed:
[src] Move curand handle out of CuRand class and into CuDevice.  (#3196)

Commit:
[build] Make MKL the default BLAS library, add installation scripts (#3194)

Dr-Desty-Nova committed:
[build] check for i686 as a valid prefix for Android triplets (#3213)

Add check for i686 as a valid prefix for Android triplets.

Kaldi configure script for 32-bit x86 expects x86 cross-compile
toolchains to start with x86.

Android compilers, however, have triplets like i686-linux-android-gcc
and i686-linux-android-ar, using i686 as a prefix.

Fixes: #3208

saikiranvalluri committed:
Added sre10 data prep script

Commit:
[build] Fix configure breakage from #3194 (MKL default)

Establish MATHLIB as the single ground-truth source for the chosen
matrix algebra library.

saikiranvalluri committed:
Fix in make_callhome.sh

teinhonglo committed:
[build] Add missing line continuation '\' in tfrnnlmbin/Makefile (#3218)

huangruizhe committed:
[src] Fix nnet2 DctComponent test failure (#3225)

kangshiyin committed:
[src] Update  CUDA code to avoid synchronization errors on compute capability 7.x (#3211)

huangruizhe committed:
[src] fix nnet2 DCTCompnent test failure -- removing anther dct_keep_dim=1 (#3226)

Commit:
[build] Remove references to deprecated MKL libs in gst_plugin (#3229)

These libraries are long gone, see
https://software.intel.com/en-us/articles/intel-math-kernel-library-intel-mkl-112-deprecations

Fixes #2202

rezame committed:
[scripts] Fix default params in nnet3 segmentation script (#3230)

KarelVesely84 committed:
[src] Correct sanity check in nnet-example-utils.cc (nnet3) (#3232)

danpovey committed:
Revert "[src] Update  CUDA code to avoid synchronization errors on compute capability 7.x (#3211)" (#3236)

This reverts commit 4cfbd217ad7e0324dba90eb31d9467dae2da4e17.

mcalahan committed:
[build] .gitignore autogenerated /tools/python/ (#3241)

When python2 binary/link doesn't exist, the extra/check_dependencies.sh
creates a symlink for python2 in tools/python/.

This location is not ignored in .gitignore so it causes git to assume
directory is dirty.

Commit:
[scripts] Enhance argument checks in nnet3/align_lats.sh (#3243)

naxingyu committed:
[egs] invoke 'python2.7' not 'python' when using mmseg (#3244)

KarelVesely84 committed:
[scripts] Make getting nnet3 model context more robust  (#3247)

- in 'check_model_contexts()' when parsing "left-context,right-context",
  don't apply 'nnet3-copy --edits=' on 'init.raw' !!!
   - the 'init.raw' is used as feature transform prior to LDA estimation,
     and here we don't need to rename the output layers.
   - renaming the output layers is relevant to multilingual networks,
     where we can have output layers 'output-0 output-1 output-2',
     layer renaming 'output-0 -> output' is used to get the contexts.
     But we cannot use the same rename on 'init.raw', which does not
     have a component with name 'output-0'.

- We also avoid calling pipe with 'nnet3-info {} | head -n 4'
   - if nnet3-info fails, the pipeline still returns zero status
     and no Python stack-trace was displayed.

zh794390558 committed:
[egs] Fix hkust_data_prep.sh w.r.t. iconv mac compatibility issue (#3250)

indra622 committed:
[egs] Update RM chain recipe with more recent configuration (#3237)

sunshines14 committed:
[egs] Make voxceleb recipe work with latest version of the dataset  (#3249)

saikiranvalluri committed:
Code fixes for final stage VB smoothing

saikiranvalluri committed:
Added ivector-extractor-copy in ivectorbin

indra622 committed:
[egs] Improve chain example script for Resource Management (RM) (#3252)

luitjens committed:
[src] GPU-related changes for speed and correctness on newer arch's. (#3221)

Changes:

cu-array-inl.h, cu-packed-matrix.cc:
   Remove unecessary synchronization.  Synchronization will occur with
   stream semantics

cu-device.h, cu-device.cc, cuda-common.h, cuda_64bit.mk
   Add a handle for cusolverDN library.  Future changes will rely on
   this.

cu-kernels-ansi.h, cu-kernels.cu, cu-kernels.h:
   Add RowSumMat kernel support which mirriors ColSumMat but operators on rows.

cu-matrix.cc:
   make cudaMemset2D asynchronous.  Synchronization is handled via
   streams.

cu-value.h:
   Added -= operator which mirriors += operator

cu-vector.cc, cu-vector.h:
   Added ApplyLogSoftMax which matches CPU version.
   Remove stream synchronization on AddMatVec (handled by streams)
   Use direct kernel for row sum instead of a mat vec.  This is more
   efficient as it avoids extra allocation and memset.

cu-sparse-matrix-test.cc:
   adjusted epislon to be more tolerant of order of operations floating
   point error.

jyhnnhyj committed:
[egs] Update voxceleb v1 preparation scripts (#3255)

nshmyrev committed:
[build] Note default=MKL; cosmetic fix (#3257)

zh794390558 committed:
[egs] Fix to hkust_data_prep.sh w.r.t. how mmseg is checked for (#3240)

KarelVesely84 committed:
[egs] In WSJ run_ivector_common.sh, expose i-vector #jobs config to run script (#3248)

* wsj: exposing i-vector #jobs config to the master script,

- there is the 3layer granularity: threads in a binary, mid-level processes
  (summing stats and saving them), top-level processes (-nj).
- in our cluster it is difficult to get a JOB with too many threads
  (the jobs get stuck in the queue).
- if the settings are exposed into the top-level script, it will be
  easier to change to setup to make it work well by modifying the top
  level script only...

alx741 committed:
[egs] Add Spanish dimex100 example (#3254)

Commit:
[build] Build and configure OpenBLAS; default to it on non-x64 machine (#3261)

* Default to OpenBLAS on non-x64 platforms.
* Download and build specific release of OpenBLAS from GitHub.
* Configure looks in the OpenBLAS build location only; the user
  can still specify `--openblas-root=`.
* Refactor switch parsing, chiefly to distinguish whether the
  `--openblas-root=` switch was not given, or was but pointed
  to a non-existing directory.
* Trim some dead ATLAS-related code left after commit 3e77220b3.
* Name all ATLAS configuration functions `*_atlas_*`.
* Move common Linux CUDA and Speex configuration out of math
  library configuration functions.

Fix: #3222
Mention: #3228

songyf committed:
[scripts] Fix of a bug in segmentation.pl (#3256)

luitjens committed:
[src] Fixes to cuda unit tests. (#3268)

hugovbraun committed:
[src] Adding GPU/CUDA lattice batched decoder + binary (#3114)

saikiranvalluri committed:
VB_diarization code mirrored in ttps://github.com/GoVivaceInc/

DongjiGao committed:
[src] Fix unit-test failure UnitTestCuMatrixSetRandn() (#3274)

huangruizhe committed:
[src,build]  Removed cusolver for now (not needed yet; caused build problems) (#3276)

saikiranvalluri committed:
David's changes phase 1

hhadian committed:
[scripts] Make fix_data_dir.sh remove utterances which have bad duration. (#3275)

Commit:
[scripts] Make generate_plots.py python3-compatible (#3280)

* [scripts] Make generate_plots.py python3-compatible

 * Use raw strings for TeX; python3 has more escapes.
 * Use sorted() for sorting, since many functions return
   transient iterators.

Also, mostly cosmetic fixes:
 * Do not print the missing imports warning before parsing
   the command line (it can use --help).
 * Suppress false-positive matplotlib warning.
 * Add metavars to help message so it reads better.
 * Make logging lines more concise (no point printing time).
 * Simplify logger initialization.
 * Use logging.warning for warnings; warnings.warn() is for
   library code only.

* fixup! [scripts] Make generate_plots.py python3-compatible

xsawyerx committed:
[scripts] Add --one-based option to split_scp.pl (#3279)

rezame committed:
[scripts] Allow UTF utterance-ids by removing unnecessary assert (#3283)

danpovey committed:
[src] Keep nnet output in the [-30,30] range required by chain denominator (#3024)

Intended to help resolve occasional instabilities.  Not 100% clear that it was helping, as instabilities don't tend to be that reproducible.

xsawyerx committed:
[scripts] Clean up filehandle usage in split_scp.pl (#3285)

* Move to three-arg open() for security reasons:

A two-arg open() call is a security flaw, since the filename could affect
the method of opening a file. This can be very, *very* risky. If the input
comes from the command line, triggered by a system that uses information from
the customer to affect filenames, it could *easily* lead to serious security
risks.

* Replace bareword filehandles with proper variables:

When using bareword filehandles, you are creating global variables. Instead,
we're using proper scalars for the filehandles. Much safer!

This commit also closes filehandles that were left open. Perl tries to
implicitly close them for you, but it's better to be explicit.

danpovey committed:
[src] Fix to bug in online-feature.cc that caused crash at end of utterance. (#3286)

xsawyerx committed:
[scripts] Use correct compile-time regex syntax in split_scp.pl (#3287)

Regular expressions should be written with // to allow the language to do
compile-time checks for the regexp pattern.

This is from `perldoc perlop`:

    If the right argument is an expression rather than a search pattern,
    substitution, or transliteration, it is interpreted as a search pattern at
    run time. Note that this means that its contents will be interpolated
    twice, so

        '\\' =~ q'\\';

    is not ok, as the regex engine will end up trying to compile the pattern
    "\", which it will consider a syntax error.

xiaohui-zhang committed:
[scripts] Fix a typo in steps/dict/learn_lexicon_bayesian.sh (#3288)

DongjiGao committed:
[egs,scripts] Scripts and an example of BPE-based sub-word decoding (#3101)

Commit:
[scripts] Add trainer option --trainer.optimization.num-jobs-step (#3205)

Commit:
[egs] Add MGB-5 recipe; https://arabicspeech.org/mgb5 (#3299)

danpovey committed:
Revert "[scripts] Clean up filehandle usage in split_scp.pl (#3285)" (#3307)

This reverts commit b17fc84472b9526b518c907148e74958f622135f.

bringtree committed:
[src] Fix bug in GeneralMatrix::Uncompress() (#3304)

KarelVesely84 committed:
[src] nnet1: lstm training, introducing cursors when slicing the training data (#3302)

Commit:
[doc] add an omission in Doxyfile (#3309)

Commit:
[scripts] Fix utils/split_scp.pl breakage (#3308)

saikiranvalluri committed:
[egs] Bug-fix to shebang in fisher_callhome_spanish (#3312)

Commit:
[scripts] Fix error messages in run.pl (#3314)

vimalmanohar committed:
[egs] New chime-5 recipe (#2893)

phanisankar-nidadavolu committed:
[scripts,egs] Made changes to the augmentation script to make it work for ASR and speaker ID (#3119)

Now multi-style training with noise and reverberation is an option (instead of speed augmentation).
Multi-style training seems to be more robust to unseen/noisy conditions.

phanisankar-nidadavolu committed:
[egs] updated local/musan.sh to steps/data/make_musan.sh in speaker id scripts (#3320)

Commit:
[src] Fix sample rounding errors in extract-segments (#3321)

With a segments file constructed from exact wave file durations
some segments came out one sample short. The reason is the
multiplication of the float sample frequency and double audio
time point is inexact. For example, float 8000.0 multiplied
by double 2.03 yields 16239.99999999999, one LSB short of the
correct sample number 16240.

Also changed all endpoint calculations so that they performed
in seconds, not sample numbers, as this does not require a
conversion in nearly every comparison, and report positions in
diagnostic messages also in seconds, not sample numbers.

Commit:
[src,scripts]Store frame_shift, utt2{dur,num_frames}, .conf with features (#3316)

Generate utt2dur and utt2num_frames during feature extraction,
and store frame period in frame_shift file in feature directory.

Copy relevant .conf files used in feature extraction into
the conf/ subdirectory with features.

Add missing validations and options in some extraction scripts.

mdoulaty committed:
[build] Initial version of Docker images for (CPU and GPU versions) (#3322)

Commit:
[scripts] fix typo/bug in make_musan.py (#3327)

phanisankar-nidadavolu committed:
[scripts] Fixed misnamed variable in data/make_musan.py (#3324)

Commit:
[scripts] Trust frame_shift and utt2num_frames if found (#3313)

Getting utt2dur involves accessing wave files, and potentially
running full pipelines in wav.scp, which may take hours for a
large data set. If utt2num_frames exists, use it instead if
frame rate is known.

Issue: #3303
Fixes: #3297 "cat: broken pipe"

Commit:
[scripts] typo fix in augmentation script (#3329)

Fixes typo in #3119

Commit:
[scripts] handle frame_shit and utt2num_frames in utils/ (#3323)

subset_data_dir.sh has been refactored thoroughly so that its
logic can be followed easier. It has been well tested and
dogfooded.

All changes here are necessary to subset, combine and verify
utt2num_frames, and copy frame_shift to new directories where
necessary.

Commit:
[scripts] Extend combine_ali_dirs.sh to combine alignment lattices (#3315)

Relevant discussion:
https://groups.google.com/forum/#!topic/kaldi-help/2uxfByEAmfw

alumae committed:
[src] Fix rare case when segment end rounding overshoots file end in extract-segments (#3331)

phanisankar-nidadavolu committed:
[scripts] Change --modify-spk-id default to False; back-compatibility fix for #3119 (#3334)

danpovey committed:
[build] Add easier configure option in failure message of configure (#3335)

Shujian2015 committed:
[scripts,minor] Fix typo in comment (#3338)

saikiranvalluri committed:
[src,egs] Add option for applying SVD on trained models (#3272)

luitjens committed:
[src] Add interfaces to nnet-batch-compute that expects device input. (#3311)

This avoids a ping pong of memory to host.

Implementation now assumes device memory.  interfaces will allocate
device memory and copy to it if data starts on host.

Add a cuda matrix copy function which clamps rows.  This is much
faster than copying one row at a time and the kernel can handle the
clamping for free.

entn-at committed:
[build] Update GCC support check for CUDA toolkit 10.1 (#3345)

naxingyu committed:
[egs] Fix to aishell1 v1 download script (#3344)

vimalmanohar committed:
[scripts] Support utf-8 files in some scripts (#3346)

huangruizhe committed:
[src] Fix potential underflow bug in MFCC, RE energy floor, thx: Zoltan Tobler (#3347)

bringtree committed:
[scripts]: add warning to nnet3/chain/train.py about ineffective options (#3341)

vimalmanohar committed:
[scripts] Fix regarding UTF handling in cleanup script (#3352)

hhadian committed:
[scripts] Change encoding to utf-8 in data augmentation scripts (#3360)

luitjens committed:
[src] Add CUDA accelerated MFCC computation. (#3348)

* Add CUDA accelerated MFCC computation.

Creates a new directory 'cudafeat' for placing cuda feature extraction
components as it is developed.  Added a directory 'cudafeatbin' for
placing binaries that are cuda accelerated that mirrior binaries
elsewhere.

This commit implements:
  feature-window-cuda.h/cu which implements a feature window on the device
    by copying it from a host feature window.
  feature-mfcc-cuda.h/cu which implements the cuda mfcc feature
    extractor.
  compute-mfcc-feats-cuda.cc which mirriors compute-mfcc-feats.cc

  There were also minor changes to other files.

* Only build cuda binaries if cuda is enabled

luitjens committed:
[src] Optimizations for batch nnet3.  The issue fixed here is that (#3351)

small cuda memory copies are inefficeint because each copy can
add multiple micro-seconds of latency.  The code as written
would copy a small matrices or vectors to and from the tasks one
after another.  To avoid this i've implemented a batched matrix
copy routine.  This takes arrays of matrix descriptions for the
input and output and batches the copies in a single kernel call.
This is used in both FormatInputs and FormatOutputs to reduce
launch latency overhead.

The kernel for the batched copy uses a trick to avoid a memory
copy of the host paramters.  The parameters are put into a struct
containing a static sized array.  These parameters are then marshalled
like normal cuda parameters.  This avoids additional launch latency
overhead.

There is still more work to do at the beginning and end of nnet3.
In particular we may want to batch the clamped memory copies and
the large number of D2D copies at the end.  I haven't fully tracked
those down and may return to them in the future.

Shujian2015 committed:
[scripts,minor] Remove outdated comment (#3361)

DatatangAI committed:
[egs] A kaldi recipe based on the corpus named "aidatatang_200zh". (#3326)

KarelVesely84 committed:
[src] nnet1: changing end-rule in 'nnet-train-multistream', (#3358)

- end the training when there is no more data to refill one of the streams,
- this avoids overtraining to the 'last' utterance,

jtrmal committed:
[scripts] Fix how the empty (faulty?) segments are handled in data-cleanup code (#3337)

danpovey committed:
[src] Fix to bug in ivector extraction causing assert failure, thx: sray (#3364)

danpovey committed:
[src] Fix to bug in ivector extraction causing assert failure, thx: sray (#3365)

hainan-xv committed:
[scripts] add script to compute dev PPL on kaldi-rnnlm (#3340)

HuangZiliAndy committed:
[scripts,egs] Small fixes to diarization scripts (#3366)

danpovey committed:
[egs] Modify split_scp.pl usage to match its updated code (#3371)

luitjens committed:
[src] Fix non-cuda `make depend` build by putting compile guards around header. (#3374)

mdoulaty committed:
[build] Docker docs update and minor changes to the Docker files  (#3377)

mahsa7823 committed:
[egs] Scripts for MATERIAL ASR (#2165)

luitjens committed:
[src] Batch nnet3 optimizations.  Batch some of the copies in and copies out (#3378)

langep committed:
[build] Widen cuda guard in cudafeat makefile. (#3379)

KarelVesely84 committed:
[scripts] nnet1: updating the scripts to support 'online-cmvn', (#3383)

luitjens committed:
[build,src] Enhancements to the cudamatrix/cudavector classes. (#3373)

* Added CuSolver to the matrix class.  This is only supported with
Cuda 9.1 or newer.  Calling CuSolver code without Cuda 9.1 or newer
will result in a runtime error.

This change required some changes to the build system which requires
versioning the configure script. This forces everyone to reconfigure.
Failure to reconfigure would result in linking and build errors on
some systems.

naxingyu committed:
[egs] Fix perl `use encoding` deprecation (#3386)

hhadian committed:
[scripts] Add max_active to align_fmllr_lats.sh to prevent rare crashes (#3387)

luitjens committed:
[src] Implemented CUDA acclerated online cmvn. (#3370)

This patch is part of a larger effort to implement the entire online feature pipeline in CUDA so that wav data is transfered to the device and never copied back to the host.
This patch includes a new binary cudafeatbin/apply-cmvn-online.cc which for the most part matches online2bin/apply-cmvn-online.
This binary is primarily for correctness testing and debugging as it makes no effort to compute multiple features in parallel on the device.
The CUDA performance is dominiated by the cost of copying the feature to and from the device. While there is a small speedup I do not expect this binary to be used in production.
Instead users will use the upcomming online-pipeline which will take features directly from the mfcc computation on the device and pass results to the next part of the pipeline.

Summary of changes:

Makefile:
   Added online2 dependencies to cudafeat, cudafeatbin, cudadecoder, and cudadecoderbin.
cudafeat\:
   Makefile:  added online2 dependency, added new .cu/.h files
   feature-online-cmvn-cuda.cu/h:  implements online-cmvn in cuda.
cudafeatbin\:
   Makefile:  added new binary, added online2 dependency
   apply-cmvn-online-cuda.cc:  binary which mimics online2bin/apply-cmvn-online

Correctness testing:

The correctness was tested by generating set of 20000 features and then running the CPU binary and GPU binary and comparing results using featbin/compare-feats.

../online2bin/apply-cmvn-online /workspace/models/LibriSpeech/ivector_extractor/global_cmvn.stats "scp:mfcc.scp" "ark,scp:cmvn.ark,cmvn.scp"
./apply-cmvn-online-cuda /workspace/models/LibriSpeech/ivector_extractor/global_cmvn.stats "scp:mfcc.scp" "ark,scp:cmvn-cuda.ark,cmvn-cuda.scp"

../featbin/compare-feats ark:cmvn-cuda.ark ark:cmvn.ark
LOG (compare-feats[5.5.1301~3-17818]:main():compare-feats.cc:105) self-product of 1st features for each column dimension:  [ 5.52221e+09 9.1134e+09 5.92818e+09 7.42173e+09 7.48633e+09 7.21316e+09 6.9515e+09 7.03883e+09 6.40267e+09 5.83088e+09 5.01438e+09 5.1575e+09 4.28688e+09 3.529e+09 3.12182e+09 2.28721e+09 1.76343e+09 1.35117e+09 8.72517e+08 5.31836e+08 2.65112e+08 9.20308e+07 1.24084e+07 3.56008e+06 4.25283e+07 1.09786e+08 1.88937e+08 2.60207e+08 3.23115e+08 3.56371e+08 3.69035e+08 3.65216e+08 3.89125e+08 4.07064e+08 3.40407e+08 2.65444e+08 2.50244e+08 2.05726e+08 1.60606e+08 1.07217e+08 ]

LOG (compare-feats[5.5.1301~3-17818]:main():compare-feats.cc:106) self-product of 2nd features for each column dimension:  [ 5.5223e+09 9.11355e+09 5.92812e+09 7.4218e+09 7.48666e+09 7.21338e+09 6.95174e+09 7.03895e+09 6.40254e+09 5.83113e+09 5.01411e+09 5.15774e+09 4.28692e+09 3.52918e+09 3.122e+09 2.28693e+09 1.76326e+09 1.3513e+09 8.72521e+08 5.31802e+08 2.65137e+08 9.20296e+07 1.2408e+07 3.5604e+06 4.25301e+07 1.09793e+08 1.88933e+08 2.60217e+08 3.23124e+08 3.56371e+08 3.69007e+08 3.65176e+08 3.89104e+08 4.07067e+08 3.40416e+08 2.65498e+08 2.50196e+08 2.057e+08 1.60612e+08 1.07192e+08 ]

LOG (compare-feats[5.5.1301~3-17818]:main():compare-feats.cc:107) cross-product for each column dimension:  [ 5.52209e+09 9.11229e+09 5.92538e+09 7.41665e+09 7.47877e+09 7.20269e+09 6.93785e+09 7.02284e+09 6.38411e+09 5.81143e+09 4.99389e+09 5.13753e+09 4.26792e+09 3.51154e+09 3.10676e+09 2.27436e+09 1.75322e+09 1.34367e+09 8.67367e+08 5.28672e+08 2.63516e+08 9.14194e+07 1.23215e+07 3.53409e+06 4.21905e+07 1.08872e+08 1.87238e+08 2.57779e+08 3.19827e+08 3.5252e+08 3.64691e+08 3.60529e+08 3.84482e+08 4.02396e+08 3.36136e+08 2.61631e+08 2.46931e+08 2.03079e+08 1.5856e+08 1.05738e+08 ]

LOG (compare-feats[5.5.1301~3-17818]:main():compare-feats.cc:111) Similarity metric for each dimension  [ 0.99997 0.999871 0.999532 0.999311 0.998968 0.998533 0.998019 0.997719 0.997111 0.996644 0.995941 0.996104 0.995572 0.995028 0.995147 0.994445 0.994258 0.994402 0.994095 0.994084 0.993934 0.993363 0.993015 0.992655 0.992037 0.991645 0.991017 0.990649 0.98981 0.989195 0.988267 0.987222 0.988093 0.98853 0.987442 0.985534 0.986858 0.987196 0.987242 0.986318 ]
 (1.0 means identical, the smaller the more different)
LOG (compare-feats[5.5.1301~3-17818]:main():compare-feats.cc:116) Overall similarity for the two feats is:0.993119 (1.0 means identical, the smaller the more different)
LOG (compare-feats[5.5.1301~3-17818]:main():compare-feats.cc:119) Processed 20960 feature files, 0 had errors.
LOG (compare-feats[5.5.1301~3-17818]:main():compare-feats.cc:126) Features are considered similar since 0.993119 >= 0.99

Shujian2015 committed:
[egs] Fixed file path RE augmentation, in aspire recipe (#3388)

vimalmanohar committed:
[scripts] Update taint_ctm_edits.py, RE utf-8 encoding (#3392)

danpovey committed:
[src] Change nnet3-am-copy to allow more manipulations (#3393)

Shujian2015 committed:
[egs] Remove confusing setting of overridden num_epochs variable in aspire (#3394)

hhadian committed:
[build] Add a missing dependency for "decoder" in Makefile (#3397)

hugovbraun committed:
[src] CUDA decoder performance patch (#3391)

danpovey committed:
[build,scripts] Dependency fix; add cross-references to scripts (#3400)

naxingyu committed:
[egs] Fix cleanup-after-partial-download bug in aishell  (#3404)

YiwenShaoStephen committed:
[src] Change functions like AppiyLog() to all work out-of-place (#3185)

rosun82 committed:
[src] Make stack trace display more user friendly (#3406)

danpovey committed:
[egs] Fix to separators in Aspire reverb recipe (#3408)

Shujian2015 committed:
[egs] Fix to separators in Aspire, related to #3408 (#3409)

KarelVesely84 committed:
[src] online2-tcp, add option to display start/end times (#3399)

luitjens committed:
[src] Remove debugging assert in cuda feature extraction code (#3411)

hhadian committed:
[scripts] Fix to checks in adjust_unk_graph.sh (#3410)

bash test `-f` does not work for `phones/` which is a directory. Changed it to `-e`.

luitjens committed:
[src] Added GPU feature extraction (will improve speed of GPU decoding) (#3390)

Currently only supports MFCC features.

luitjens committed:
[src] Fix build error introducted by race condition in PR requests/accepts. (#3412)

luitjens committed:
[src] Added error string to CUDA allocation errors. (#3413)

LvHang committed:
[src] Fix CUDA_VSERION number in preprocessor checks (#3414)

luitjens committed:
[src] Fix build of online feature extraction with older CUDA version (#3415)

LvHang committed:
[src] Update Insert function of hashlist and decoders (#3402)

makes interface of HashList more standard; slight speed improvement.

LvHang committed:
[src] Fix spelling mistake in #3415 (#3416)

LvHang committed:
[build] Fix configure bug RE CuSolver (#3417)

luitjens committed:
[src] Enable an option to use the GPU for feature extraction in GPU decoding (#3420)

This is turned on by using the option
--gpu-feature-extract=true.  By default this is on.  We provie the
option to turn it off because in situations where CPU resources
are unconfined you can get slightly higher performance with CPU
feature extraction but in most cases GPU feature extraction is faster
and has more stable performance.  In addition a user may wish to
turn it off to support models where feature extraction is currently
incomplete (e.g. FBANK, PLP, PITCH, etc).  We will add those
features in the future but for now a user wanted to decode those
models should place feature extraction on the host.

rezame committed:
[egs] Replace $cuda_cmd with $train_cmd for FarsDat (#3426)

cloudhan committed:
[src] Remove outdated comment (#3148) (#3422)

hugovbraun committed:
[src] Adding missing thread.join in CUDA decoder and fixing two todos (#3428)

danpovey committed:
[build] Add missing lib dependency in cudafeatbin (#3427)

Shujian2015 committed:
[egs] Small fix to aspire run_tdnn_7b.sh (#3429)

danpovey committed:
[build] Fix to cuda makefiles, thanks: yiyidhuang@gmail.com (#3431)

danpovey committed:
[build] Add missing deps to cuda makefiles, thanks: yiyidhuang@gmail.com (#3432)

boystray committed:
[egs] Fix encoding issues in Chinese ASR recipe (#3430) (#3434)

danpovey committed:
Revert "[src] Update Insert function of hashlist and decoders (#3402)" (#3436)

This reverts commit 5cc7ce063aaa96800c38c25c6b9fa8bf455578c9.

LvHang committed:
[src] Update Insert function of hashlist and decoders (#3402) (#3438)

makes interface of HashList more standard; slight speed improvement.  Fixed version of #3402

rayworks committed:
[build] Fix the cross-compiling issue for Android under MacOS (#3435)

luitjens committed:
[src] Marking operator as __host__ __device__ to avoid build issues (#3441)

avoids cudafeat build failures with some CUDA toolkit versions

naxingyu committed:
[egs] Fix perl encoding bug (was causing crashes) (#3442)

hugovbraun committed:
[src] Cuda decoder fixes, efficiency improvements (#3443)

naxingyu committed:
[scripts] Fix shebang of taint_ctm_edits.py to invoke python3 directly (#3445)

danpovey committed:
[src] Fix to a check in nnet-compute code (#3447)

csukuangfj committed:
[src,scripts] Various typo fixes and stylistic fixes (#3153)

saikiranvalluri committed:
Merge branch 'master' into VariationalBayes_SpeakerDiarization
